runtime/lib/ttmetal/runtime.cpp

```diff
@@ -1,3 +1,182 @@
+  desc->stride = stride;
+  desc->itemsize = itemsize;
+  desc->dataType = dataType;
+  return Tensor(static_pointer_cast<void>(desc), data);
+}
+
+Device openDevice(std::vector<int> const &deviceIds,
+                  std::vector<std::uint8_t> const &numHWCQs) {
+  assert(numHWCQs.empty() || numHWCQs.size() == deviceIds.size());
+  std::shared_ptr<DeviceMesh> deviceMesh = std::make_shared<DeviceMesh>();
+  int i = 0;
+  for (int deviceId : deviceIds) {
+    uint8_t num_hw_cqs = numHWCQs.empty() ? 1 : numHWCQs[i];
+    deviceMesh->push_back(CreateDevice(deviceId, num_hw_cqs));
+    ++i;
+  }
+  return static_pointer_cast<void>(deviceMesh);
+}
+
+void closeDevice(Device device) {
+  DeviceMesh &deviceMesh = device.as<DeviceMesh>();
+  for (::tt::tt_metal::Device *device : deviceMesh) {
+    ::tt::tt_metal::CloseDevice(device);
+  }
+}
+
+Event submit(Device deviceHandle, Binary executableHandle,
+             std::uint32_t programIndex,
+             std::vector<Tensor> const &inputHandles,
+             std::vector<Tensor> const &outputHandles) {
+  ::tt::target::metal::TTMetalBinary const &fbb = *getBinary(executableHandle);
+  ::tt::target::metal::Program const *program =
+      fbb.programs()->Get(programIndex);
+  std::vector<std::pair<std::uint32_t, std::shared_ptr<::tt::tt_metal::Buffer>>>
+      inputs;
+  inputs.reserve(inputHandles.size());
+  assert(inputHandles.size() == program->inputs()->size() &&
+         "Input size mismatch");
+  for (unsigned i = 0; i < inputHandles.size(); ++i) {
+    inputs.emplace_back(
+        program->inputs()->Get(i)->global_id(),
+        static_pointer_cast<::tt::tt_metal::Buffer>(inputHandles[i].handle));
+  }
+
+  std::vector<std::pair<std::uint32_t, std::shared_ptr<::tt::tt_metal::Buffer>>>
+      outputs;
+  outputs.reserve(outputHandles.size());
+  assert(outputHandles.size() == program->outputs()->size() &&
+         "Output size mismatch");
+  for (unsigned i = 0; i < outputHandles.size(); ++i) {
+    outputs.emplace_back(
+        program->outputs()->Get(i)->global_id(),
+        static_pointer_cast<::tt::tt_metal::Buffer>(outputHandles[i].handle));
+  }
+
+  DeviceMesh &deviceMesh = deviceHandle.as<DeviceMesh>();
+  std::shared_ptr<Events> events = std::make_shared<Events>();
+  std::size_t cq_id = 0;
+  assert(program->device_programs()->size() == deviceMesh.size() &&
+         "Device programs size mismatch");
+  for (std::size_t i = 0; i < program->device_programs()->size(); ++i) {
+    ::tt::tt_metal::Device *device = deviceMesh[i];
+    ::tt::target::metal::DeviceProgram const *deviceProgram =
+        program->device_programs()->Get(i);
+    for (::tt::target::metal::CommandQueue const *cq :
+         *deviceProgram->command_queues()) {
+      events->push_back(
+          executeCommandQueue(device, cq, cq_id, inputs, outputs));
+      ++cq_id;
+    }
+  }
+
+  return static_pointer_cast<void>(events);
+}
+
+void wait(Event event) {
+  Events events = event.as<Events>();
```

Comment:
I'm probably missing something here, but is this intentional (converting from what looks like singular Event, to vector of metal Events)?

Reply:
Yes, notice above that line 156 creates a vector of metal events.  This is because there might be many underlying events created and tracked from this program `submit` (i.e. executing on multi-device / multi-command queue) that all need to be wrapped up in a single event.  To the FE runtime it should only care about managing barriers at program submit granularity.